derive_client/_bridge/client.py

```diff
@@ -324,173 +358,163 @@ def deposit_drv(self, amount: int) -> BridgeTxResult:
+        return wait_for_event(context.target_w3, filter_params, condition=matching_message_id)
 
     def poll_bridge_progress(self, tx_result: BridgeTxResult) -> BridgeTxResult:
-        # TODO: handle non-pending status
-
-        source_tx = tx_result.source_tx
-        target_tx = tx_result.target_tx
-
-        source_w3 = get_w3_connection(tx_result.source_chain)
-        target_w3 = get_w3_connection(tx_result.target_chain)
-
-        # 1. Timeout during source_tx.tx_receipt
-        if not source_tx.tx_receipt:
-            print(f"⏳ Checking source chain [{tx_result.source_chain.name}] tx receipt for {source_tx.tx_hash}")
-            source_tx.tx_receipt = wait_for_tx_receipt(w3=source_w3, tx_hash=source_tx.tx_hash)
-
-        # 2. Timeout waiting for event_log on target chain
-        if not target_tx.tx_hash:
-            match tx_result.bridge:
-                case BridgeType.SOCKET:
-                    event_log = self.fetch_socket_event_log(tx_result)
-                case BridgeType.LAYERZERO:
-                    event_log = self.fetch_lz_event_log(tx_result)
-                case _:
-                    raise ValueError()
-            target_tx.tx_hash = event_log["transactionHash"].to_0x_hex()
+        if tx_result.status is not TxStatus.PENDING:
+            raise AlreadyFinalizedError(f"Bridge already in final state: {tx_result.status.name}")
+
+        bridge_event_fetchers = {
+            BridgeType.SOCKET: self.fetch_socket_event_log,
+            BridgeType.LAYERZERO: self.fetch_lz_event_log,
+        }
+        if (fetch_event := bridge_event_fetchers.get(tx_result.bridge)) is None:
+            raise ValueError(f"Invalid bridge_type: {tx_result.bridge}")
+
+        direction = "withdraw" if tx_result.source_chain == ChainID.DERIVE else "deposit"
+        context = self._make_bridge_context(
+            direction=direction,
+            bridge_type=tx_result.bridge,
+            currency=tx_result.currency,
+        )
+
+        # 1. TimeoutError as exception during source_tx.tx_receipt
+        if not tx_result.source_tx.tx_receipt:
+            print(
+                f"⏳ Checking source chain [{tx_result.source_chain.name}] tx receipt for {tx_result.source_tx.tx_hash}"
+            )
+            tx_result.source_tx.exception = None
+            try:
+                tx_result.source_tx.tx_receipt = wait_for_tx_receipt(
+                    w3=context.source_w3, tx_hash=tx_result.source_tx.tx_hash
+                )
+            except Exception as e:
+                tx_result.source_tx.exception = e
+
+        # 2. target_tx is None (i.e. TimeoutError when waiting for event_log on target chain)
+        if not tx_result.target_tx:
+            try:
+                event_log = fetch_event(tx_result, context)
+                tx_result.target_tx = TxResult(event_log["transactionHash"].to_0x_hex())
+            except TimeoutError:
+                return tx_result
```

Comment:
if we don't have a `tx_result.target_tx`, that means either we enter for first time (if called from internally here) or otherwise we must have had some `TimeoutError` in `wait_for_event`. Any other exception (i.e. `BridgeEventParseError` in parsing `source_tx.tx_receipt`, or any non-`TimeoutError` in `wait_for_event` will raise).

This is the one exception we cannot properly store, since it does not belong to `source_tx.exception` nor `target_tx.exception`